f/common_logic/data_validation.py

```diff
@@ -0,0 +1,152 @@
+import csv
+import json
+import logging
+from pathlib import Path
+
+import pandas as pd
+from lxml import etree
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+def detect_structured_data_type(file_path: str) -> str:
+    """
+    Detects the type of a structured data file, focusing on geospatial and tabular formats.
+
+    Only files with recognized extensions are considered. The function attempts to
+    parse the file to confirm it matches the expected structure. If the file type is
+    unsupported or the content does not validate, 'unsupported' is returned.
+
+    This function is intended to identify structured datasets used in environmental,
+    geographic, and tabular data pipelines. It supports:
+    - Geospatial formats: geojson, kml, gpx
```

Comment:
in the same way you use pandas to read tabular files, would we consider using [fiona](https://pypi.org/project/fiona/) as a general-purpose reader for vector spatial files? (It will also handle ESRI for you)

I'm going to be suuuuper nitpicky and say that these are all "spatial" files, but not necessarily "geospatial".  (The latter implies a coordinate reference system is specified.)

Reply:
I think you should ignore my suggestion about Fiona for the time being.  Maybe add a code comment at the top of the file explaining your philosophy. But it's absolutely not worth code changes at this early point.